### PR TITLE
Support async generators

### LINKED ISSUE #1990: ES2018 Feature: Async Iteration (for-await-of)
## Syntax:

A variation of the for-of iteration statement which iterates over async iterable objects (async generator). An example usage would be:

```js
for await (const line of readLines(filePath)) {
  console.log(line);
}

async function* generator() { }
```

Async for-of statements are only allowed within async functions:

```diff
 IterationStatement[Yield, Await, Return]:
     ...
+    [+Await] for await ( [lookahead ≠ let] LeftHandSideExpression[?Yield, ?Await] of AssignmentExpression[+In, ?Yield, ?Await] ) Statement[?Yield, ?Await, ?Return]
+    [+Await] for await ( var ForBinding[?Yield, ?Await] of AssignmentExpression[+In, ?Yield, ?Await] ) Statement[?Yield, ?Await, ?Return]
+    [+Await] for await ( ForDeclaration[?Yield, ?Await] of AssignmentExpression[+In, ?Yield, ?Await] ) Statement[?Yield, ?Await, ?Return]

And async generator functions / methods are introduced:

+ AsyncGeneratorMethod[Yield, Await]:
+     async *PropertyName[?Yield, ?Await] ( UniqueFormalParameters[+Yield, +Await] ) { AsyncGeneratorBody }
+ 
+ AsyncGeneratorDeclaration[Yield, Await, Default]:
+     async function *BindingIdentifier[?Yield, ?Await] ( FormalParameters[+Yield, +Await] ) { AsyncGeneratorBody }
+     async function * ( FormalParameters[+Yield, +Await] ) { AsyncGeneratorBody }
+ 
+ AsyncGeneratorExpression:
+     async function * BindingIdentifier[+Yield, +Await]? ( FormalParameters[+Yield, +Await] ) { AsyncGeneratorBody }
+ 
+ AsyncGeneratorBody:
+     FunctionBody[+Yield, +Await]
```

## Spec:

TC39:
https://github.com/tc39/proposal-async-iteration
https://tc39.es/proposal-async-iteration/

ESTree:
https://github.com/estree/estree/blob/master/es2018.md

---

```js
extend interface ForOfStatement {
  await: boolean;
}
```

---

## Additional considerations

* Supported in Chrome, Safari. (Firefox?)
* Supported in Node, jsc

## Remaining Tasks:

- [ ] Update ForOfStatement node and docs (nodes.ts, docs/syntax-tree-format.md)
- [ ] Update Parser for for-await-of (`parser.ts` in `parseForStatement`)
- [ ] Update Parser for async generators (`parser.ts` allow for `AsyncFunctionExpression`, `AsyncFunctionDeclaration` which is a generator)
- [ ] Provide Unit Tests and Coverage
- [ ] Update test262 (~test262.git#2ee3864136) / test262-stream (~1.3.0)

## Test Cases

Valid:

```js
async function f() { for await (let x of []) {} }
async function *g() { for await (let x of []) {} }
o = { async method() { for await (let x of []) {} } }
o = { async *method() { for await (let x of []) {} } }
```

Invalid:

```js
function f() { for await (let x of []) {} }
function *g() { for await (let x of []) {} }
o = { method() { for await (let x of []) {} } }
o = { *method() { for await (let x of []) {} } }
```

### PR DESCRIPTION
**Pull request by lahma**
Async generators are valid constructs and went through issues parsing them. Couple issues surfaced on class handling side but unrelated to async generators per se.

**Before**

 ✔ 54265 valid programs parsed without error
 ✔ 6748 invalid programs produced a parsing error
 ✔ 1101 invalid programs did not produce a parsing error (and allowed by the whitelist file)
 ✔ 16288 valid programs produced a parsing error (and allowed by the whitelist file)
 
**After**

 ✔ 60262 valid programs parsed without error
 ✔ 6747 invalid programs produced a parsing error
 ✔ 1102 invalid programs did not produce a parsing error (and allowed by the whitelist file)
 ✔ 10291 valid programs produced a parsing error (and allowed by the whitelist file)
 
fixes #1990

### CODE REVIEW

**Comment by lahma** [comment]
Hi @ariya ! Here's a PR to allow parsing about 6000 more scripts that were unparseable before, basically allowing async generators and couple more error checks.

**Comment by ariya** [comment]
Great work @lahma, thanks! I will take a look.

File: src/messages.ts [hunk 1]
```diff
@@ -49,7 +51,6 @@ export const Messages = {
     StrictLHSPrefix: 'Prefix increment/decrement may not have eval or arguments operand in strict mode',
     StrictModeWith: 'Strict mode code may not include a with statement',
     StrictOctalLiteral: 'Octal literals are not allowed in strict mode.',
-    StrictParamDupe: 'Strict mode function may not have duplicate parameter names',
```
Comment (lahma) [inline src/messages.ts:52]: Removed this in favor of `DuplicateParameter` which aligns with V8

File: src/parser.ts [hunk 2]
```diff
@@ -20,13 +20,14 @@ interface Context {
     allowIn: boolean;
     allowStrictDirective: boolean;
     allowYield: boolean;
-    await: boolean;
+    isAsync: boolean;
```
Comment (lahma) [inline src/parser.ts:23]: Renamed this as await is keyword and isAsync describes the context

File: src/parser.ts [hunk 3]
```diff
@@ -256,6 +258,12 @@ export class Parser {
         this.errorHandler.tolerate(this.unexpectedTokenError(token, message));
     }
 
+    tolerateInvalidLoopStatement() {
```
Comment (lahma) [inline src/parser.ts:261]: Had to create separate function as for statement parsing started to fail due to cyclomatic complexity

File: src/parser.ts [hunk 4]
```diff
@@ -613,7 +621,7 @@ export class Parser {
 
         switch (this.lookahead.type) {
             case Token.Identifier:
-                if ((this.context.isModule || this.context.await) && this.lookahead.value === 'await') {
+                if ((this.context.isModule || this.context.isAsync) && this.lookahead.value === 'await') {
```
Comment (lahma) [inline src/parser.ts:624]: I think this now reads better, isModule, isAsync

File: src/parser.ts [hunk 5]
```diff
@@ -785,19 +792,20 @@ export class Parser {
         return this.finalize(node, new Node.FunctionExpression(null, params.params, method, isGenerator));
     }
 
-    parsePropertyMethodAsyncFunction(): Node.FunctionExpression {
+    parsePropertyMethodAsyncFunction(isGenerator: boolean): Node.FunctionExpression {
```
Comment (lahma) [inline src/parser.ts:795]: Might make sense at some point to put isGenerator into context

File: src/parser.ts [hunk 6]
```diff
@@ -2125,7 +2146,7 @@ export class Parser {
         return this.finalize(node, new Node.Property('init', key, computed, value, method, shorthand));
     }
 
-    parseRestProperty(params, kind): Node.RestElement {
+    parseRestProperty(params): Node.RestElement {
```
Comment (lahma) [inline src/parser.ts:2149]: Removed the unused parameter as style checks gave warning and couldn't see a reason to keep

File: src/parser.ts [hunk 7]
```diff
@@ -3018,6 +3043,12 @@ export class Parser {
         }
         this.expect(')');
 
+        if (options.hasDuplicateParameterNames) {
```
Comment (lahma) [inline src/parser.ts:3046]: Moved checks to single place where can validate against context

File: src/parser.ts (hunk 5 above)
Comment (ariya) [reply src/parser.ts:795]: This is [Boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap), though we already have some cases like that in the same file.

**Comment by lahma** [comment]
@ariya friendly ping from here 🙂 it would also be great if you have time to check https://github.com/jquery/esprima/pull/2082 , the new fork of Esprima is running far ahead already..

File: src/parser.ts (hunk 5 above)
Comment (lahma) [reply src/parser.ts:795]: So should I put this into `context` as part of this PR or what changes required?

**Comment by ariya** [comment]
Thank you @lahma 👍

**Comment by lahma** [comment]
You're most welcome, thanks for merging!

**Comment by jogibear9988** [comment]
@lahma i've merged your changes also in my fork

File: src/parser.ts [hunk 8]
```diff
@@ -1342,6 +1356,10 @@ export class Parser {
             expr = this.inheritCoverGrammar(this.matchKeyword('new') ? this.parseNewExpression : this.parsePrimaryExpression);
         }
 
+        if (isSuper && this.match('(') && !this.context.inClassConstructor) {
```
Comment (jogibear9988) [inline src/parser.ts:1359]: @lahma:
this condition is not correct when the class is not inherited

Comment (jogibear9988) [reply src/parser.ts:1359]: In esprima next there is also context.allowSuper. 
I got it from this pull req: https://github.com/jquery/esprima/pull/2047

**Pull request merged**